packages/flutter/lib/src/foundation/basic_types.dart

```diff
@@ -52,3 +54,149 @@ class BitField<T extends dynamic> {
+  /// Creates a CachingIterable using the given [Iterator] as the
+  /// source of data. The iterator must be non-null and must not throw
+  /// exceptions.
+  ///
+  /// Since the argument is an [Iterator], not an [Iterable], it is
+  /// guaranteed that the underlying data set will only be walked
+  /// once. If you have an [Iterable], you can pass its [iterator]
+  /// field as the argument to this constructor.
+  ///
+  /// You can use a `sync*` function with this as follows:
+  ///
+  /// ```dart
+  /// Iterable<int> range(int start, int end) sync* {
+  ///   for (int index = start; index <= end; index += 1)
+  ///     yield index;
+  ///  }
+  ///
+  /// Iterable<int> i = new CachingIterable<int>(range(1, 5).iterator);
+  /// print(i.length); // walks the list
+  /// print(i.length); // efficient
+  /// ```
+  CachingIterable(this._prefillIterator);
+
+  final Iterator<E> _prefillIterator;
+  final List<E> _results = <E>[];
+
+  @override
+  Iterator<E> get iterator {
+    return new _LazyListIterator<E>(this);
+  }
+
+  @override
+  Iterable<dynamic/*=T*/> map/*<T>*/(/*=T*/ f(E e)) {
+    return new CachingIterable<dynamic/*=T*/>(super.map/*<T>*/(f).iterator);
+  }
+
+  @override
+  Iterable<E> where(bool f(E element)) {
+    return new CachingIterable<E>(super.where(f).iterator);
+  }
+
+  @override
+  Iterable<dynamic/*=T*/> expand/*<T>*/(Iterable/*<T>*/ f(E element)) {
+    return new CachingIterable<dynamic/*=T*/>(super.expand/*<T>*/(f).iterator);
+  }
+
+  @override
+  Iterable<E> take(int count) {
+    return new CachingIterable<E>(super.take(count).iterator);
+  }
+
+  @override
+  Iterable<E> takeWhile(bool test(E value)) {
+    return new CachingIterable<E>(super.takeWhile(test).iterator);
+  }
+
+  @override
+  Iterable<E> skip(int count) {
+    return new CachingIterable<E>(super.skip(count).iterator);
+  }
+
+  @override
+  Iterable<E> skipWhile(bool test(E value)) {
+    return new CachingIterable<E>(super.skipWhile(test).iterator);
+  }
+
+  @override
+  int get length {
+    _precacheEntireList();
+    return _results.length;
+  }
+
+  @override
+  List<E> toList({ bool growable: true }) {
+    _precacheEntireList();
+    return new List<E>.from(_results, growable: growable);
+  }
+
+  void _precacheEntireList() {
+    while (_fillNext()) { }
+  }
+
+  bool _fillNext() {
+    final bool result = _prefillIterator.moveNext();
+    if (!result)
+      return false;
+    _results.add(_prefillIterator.current);
+    return true;
+  }
+}
+
+class _LazyListIterator<E> implements Iterator<E> {
+  _LazyListIterator(this._owner) : _index = -1;
+
+  final CachingIterable<E> _owner;
+  int _index;
+
+  @override
+  E get current {
+    assert(_index >= 0); // called "current" before "moveNext()"
+    if (_index < 0 || _index == _owner._results.length)
+      return null;
+    return _owner._results[_index];
+  }
+
+  @override
+  bool moveNext() {
+    if (_index < _owner._results.length)
```

Comment:
nit: I would have flipped this condition to put the return on the "if" branch and get rid of the "else" branch by moving the code to the main line of execution.